test/e2e_node/dockershim_checkpoint_test.go

```diff
@@ -0,0 +1,231 @@
+				framework.ExpectNoError(err, "Failed to run command %v: %v", dockerStopCmd, err)
+				dockerRmCmd := append([]string{"docker", "rm"}, ids...)
+				_, err = exec.Command("sudo", dockerRmCmd...).CombinedOutput()
+				framework.ExpectNoError(err, "Failed to run command %v: %v", dockerRmCmd, err)
+			})
+		})
+	})
+
+	Context("When checkpoint file is corrupted", func() {
+		It("should complete pod sandbox clean up", func() {
+			podName := "pod-checkpoint-corrupted"
+			runPodCheckpointTest(f, podName, func() {
+				By("Corrupt checkpoint file")
+				checkpoints := findCheckpoints(podName)
+				if len(checkpoints) == 0 {
+					framework.Failf("No checkpoint for the pod was found")
+				}
+				for _, file := range checkpoints {
+					f, err := os.OpenFile(file, os.O_WRONLY|os.O_APPEND, 0644)
+					framework.ExpectNoError(err, "Failed to open file %q", file)
+					_, err = f.WriteString("blabblab")
+					framework.ExpectNoError(err, "Failed to write to file %q", file)
+					f.Sync()
+					f.Close()
+				}
+			})
+		})
+	})
+})
+
+func runPodCheckpointTest(f *framework.Framework, podName string, twist func()) {
+	By(fmt.Sprintf("Creating test pod: %s", podName))
+	f.PodClient().CreateSync(&v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{Name: podName},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Image: "gcr.io/google_containers/busybox:1.24",
+					Name:  "checkpoint-test-container",
+					Command: []string{
+						"sh",
+						"-c",
+						"sleep 36000",
+					},
+				},
+			},
+		},
+	})
+
+	By("Performing disruptive operations")
+	twist()
+
+	By(fmt.Sprintf("Sanity check on pod %q", podName))
+	f.WaitForPodRunning(podName)
+
+	By("Remove test pod")
+	f.PodClient().DeleteSync(podName, &metav1.DeleteOptions{}, framework.DefaultPodDeletionTimeout)
+
+	By("Waiting for checkpoint to be removed")
+	if err := wait.PollImmediate(10*time.Second, gcTimeout, func() (bool, error) {
+		checkpoints := findCheckpoints(podName)
+		if len(checkpoints) == 0 {
+			return true, nil
+		}
+		framework.Logf("Checkpoint of %q still exists: %v", podName, checkpoints)
+		return false, nil
+	}); err != nil {
+		framework.Failf("Failed to observe checkpoint being removed within timeout: %v", err)
+	}
+}
+
+// cleanString cleans up any trailing spaces and new line character for the input string
+func cleanString(output string) string {
+	processed := strings.TrimSpace(string(output))
+	regex := regexp.MustCompile(`\r?\n`)
+	processed = regex.ReplaceAllString(processed, "")
+	return processed
+}
+
+func writeFileAndSync(path string, data []byte) error {
+	f, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
+	if err != nil {
+		return err
+	}
+	_, err = f.Write(data)
+	if err != nil {
+		return err
+	}
+	f.Sync()
+	if err1 := f.Close(); err == nil {
+		err = err1
+	}
+	return err
+}
+
+// findCheckpoints returns all checkpoint files containing input string
+func findCheckpoints(match string) []string {
```

Comment:
Due to whatever reason, kubelet may need to recreate the pod sandbox. 2 checkpoints will be presented for the pod. And there is no way we could retrieve the pod sandbox ID here, and we can only use pod name to grep the checkpoint. So it may end up with multiple checkpoints.